br/pkg/storage/locking.go

```diff
@@ -67,58 +174,136 @@ func readLockMeta(ctx context.Context, storage ExternalStorage, path string) (Lo
-	defer func() {
-		log.Info("Trying lock remote file.", zap.String("path", path), zap.String("hint", hint), logutil.ShortError(err))
-	}()
-	exists, err := storage.FileExists(ctx, path)
-	if err != nil {
-		return errors.Annotatef(err, "failed to check lock file %s exists", path)
-	}
-	if exists {
-		meta, err := readLockMeta(ctx, storage, path)
-		if err != nil {
-			return err
-		}
-		return ErrLocked{Meta: meta}
+func TryLockRemote(ctx context.Context, storage ExternalStorage, path, hint string) (lock RemoteLock, err error) {
+	writer := ExclusiveWrite{
+		Target: path,
+		Content: func(txnID uuid.UUID) []byte {
+			meta := MakeLockMeta(hint)
+			meta.TxnID = txnID[:]
+			res, err := json.Marshal(meta)
+			if err != nil {
+				log.Panic(
+					"Unreachable: a trivial object cannot be marshaled to JSON.",
+					zap.String("path", path),
+					logutil.ShortError(err),
+				)
+			}
+			return res
+		},
 	}
 
-	meta := MakeLockMeta(hint)
-	return putLockMeta(ctx, storage, path, meta)
+	lock.storage = storage
+	lock.path = path
+	lock.txnID, err = writer.CommitTo(ctx, storage)
+	if err != nil {
+		err = errors.Annotatef(err, "there is something about the lock: %s", tryFetchRemoteLock(ctx, storage, path))
+	}
+	return
 }
 
 // UnlockRemote removes the lock file at the specified path.
 // Removing that file will release the lock.
-func UnlockRemote(ctx context.Context, storage ExternalStorage, path string) error {
-	meta, err := readLockMeta(ctx, storage, path)
+func (l RemoteLock) Unlock(ctx context.Context) error {
+	meta, err := readLockMeta(ctx, l.storage, l.path)
 	if err != nil {
 		return err
 	}
 	// NOTE: this is for debug usage. For now, there isn't an Compare-And-Swap
 	// operation in our ExternalStorage abstraction.
 	// So, once our lock has been overwritten or we are overwriting other's lock,
 	// this information will be useful for troubleshooting.
-	log.Info("Releasing lock.", zap.Stringer("meta", meta), zap.String("path", path))
-	err = storage.DeleteFile(ctx, path)
+	if !bytes.Equal(l.txnID[:], meta.TxnID) {
+		return errors.Errorf("Txn ID mismatch: remote is %v, our is %v", meta.TxnID, l.txnID)
+	}
+
+	log.Info("Releasing lock.", zap.Stringer("meta", meta), zap.String("path", l.path))
+	err = l.storage.DeleteFile(ctx, l.path)
 	if err != nil {
-		return errors.Annotatef(err, "failed to delete lock file %s", path)
+		return errors.Annotatef(err, "failed to delete lock file %s", l.path)
 	}
 	return nil
 }
+
+func TryLockRemoteWrite(ctx context.Context, storage ExternalStorage, path, hint string) (lock RemoteLock, err error) {
+	target := fmt.Sprintf("%s.WRIT", path)
+	writer := ExclusiveWrite{
+		Target: target,
+		Content: func(txnID uuid.UUID) []byte {
+			meta := MakeLockMeta(hint)
+			meta.TxnID = txnID[:]
+			res, err := json.Marshal(meta)
+			if err != nil {
+				log.Panic(
+					"Unreachable: a trivial object cannot be marshaled to JSON.",
```

Comment:
nit, what does trivial at here mean?

Reply:
Some how like [`Trivial types`](https://en.cppreference.com/w/cpp/named_req/TrivialType) in C++.